gapic-common/lib/gapic/common/retry_policy.rb

```diff
@@ -0,0 +1,196 @@
+          raise ArgumentError, "A block must be provided for custom retry logic."
+        end
+        @retry_block = retry_block
+
+        self
+      end
+
+      ##
+      # Perform delay if and only if retriable.
+      #
+      # Positional argument `error` takes precedence over keyword arguments, meaning the latter
+      # will only be used iff `error` is not present.
+      #
+      # @return [Boolean] Whether the delay was executed.
+      #
+      def call error = nil, **kwargs
+        should_retry = error.nil? ? retry?(**kwargs) : retry_error?(error)
+        return false unless should_retry
+        perform_delay!
+      end
+      alias perform_delay call
+
+      ##
+      # Perform delay.
+      #
+      # @return [Boolean] Whether the delay was executed.
+      #
+      def perform_delay!
+        delay!
+        increment_delay!
+        true
+      end
+
+      ##
+      # Current delay value in seconds.
+      #
+      # @return [Numeric] Time delay in seconds.
+      #
+      def delay
+        @delay || initial_delay
+      end
+
+      ##
+      # @private
+      # @return [Boolean] Whether this error should be retried.
+      #
+      def retry_error? error
+        (defined?(::GRPC) && error.is_a?(::GRPC::BadStatus) && retry_codes.include?(error.code)) ||
+          (error.respond_to?(:response_status) &&
+            retry_codes.include?(ErrorCodes.grpc_error_for(error.response_status)))
+      end
+
+      ##
+      # @private
+      # Apply default values to the policy object. This does not replace user-provided values, it only overrides empty
+      # values.
+      #
+      # @param retry_policy [Hash] The policy for error retry. Keys must match the arguments for
+      #   {Gapic::Common::RetryPolicy.new}.
+      #
+      def apply_defaults retry_policy
+        return unless retry_policy.is_a? Hash
+        @retry_codes   ||= convert_codes retry_policy[:retry_codes]
+        @initial_delay ||= retry_policy[:initial_delay]
+        @multiplier    ||= retry_policy[:multiplier]
+        @max_delay     ||= retry_policy[:max_delay]
+
+        self
+      end
+
+
+      # @private Equality test
+      def eql? other
+        other.is_a?(RetryPolicy) &&
+          other.initial_delay == initial_delay &&
+          other.max_delay == max_delay &&
+          other.multiplier == multiplier &&
+          other.retry_codes == retry_codes &&
+          other.timeout == timeout
+      end
+      alias == eql?
+
+      # @private Hash code
+      def hash
+        [initial_delay, max_delay, multiplier, retry_codes, timeout].hash
+      end
+
+
+      private
+
+      def retry? **kwargs
+        @retry_block.nil? ? retry_with_deadline? : @retry_block.call(**kwargs)
+      end
+
+      def delay!
+        Kernel.sleep delay
+      end
+
+      def increment_delay!
+        @delay = [delay * multiplier, max_delay].min
+      end
+
+      # @private Timeout block
+      def deadline
+        @deadline ||= Time.now + timeout
```

Comment:
Here's a hint that's not as well known in Ruby as it should be. If you need the *actual* real clock time, you can use `Time.now`, but if you're just measuring how long something has taken, it's better to use `Process.clock_gettime(Process::CLOCK_MONOTONIC)`. This is because `Time.now` is sensitive to changes to the system clock (such as an update from a time server, or someone manually messing with it from the operating system), whereas the monotonic clock is guaranteed not to be affected by those factors.

Reply:
Thank you! I updated this and changed affected tests.